lib/mix/lib/mix/task.ex

```diff
@@ -445,32 +474,51 @@ defmodule Mix.Task do
     end
   end
 
-  defp run_alias([h | t], alias_args, proj, original_task, _res) when is_binary(h) do
+  defp run_alias([h | t], alias_args, proj, original_task, apps, _res) when is_binary(h) do
     case OptionParser.split(h) do
       [^original_task | args] ->
-        res = run_task(proj, original_task, args ++ alias_args)
-        run_alias(t, [], proj, original_task, res)
+        res = run_task(proj, original_task, args ++ alias_args, apps)
+        run_alias(t, [], proj, original_task, apps, res)
 
       [task | args] ->
-        res = run(task, join_args(args, alias_args, t))
-        run_alias(t, alias_args, proj, original_task, res)
+        res = do_run(task, join_args(args, alias_args, t), apps)
+        run_alias(t, alias_args, proj, original_task, apps, res)
     end
   end
 
-  defp run_alias([h | t], alias_args, proj, original_task, _res) when is_function(h, 1) do
-    res = h.(join_args([], alias_args, t))
-    run_alias(t, alias_args, proj, original_task, res)
+  defp run_alias([h | t], alias_args, proj, original_task, apps, _res)
+       when is_function(h, 1) do
+    res =
+      cond do
+        apps && Mix.Project.umbrella?() ->
+          Mix.ProjectStack.recur(fn ->
+            recur(fn _ -> h.(join_args([], alias_args, t)) end, apps)
+          end)
+
+        apps ->
+          Mix.shell().info([
+            :yellow,
+            "warning: running #{inspect(original_task)} at root level because this is not an umbrella project"
+          ])
```

Comment:
I was thinking we should refactor this warning into a private function... but then I am wondering, maybe this should be an error?

Reply:
The advantage of making it an error is that we can check it once at the top of `run_in_apps` and just raise. In here we can assume if there are `apps`, it is an umbrella.